### PR TITLE
add range tools state

### PR DESCRIPTION
**Pull request by wadjih-bencheikh18**
closes https://github.com/cheminfo/nmrium/issues/1721

### CODE REVIEW

File: src/component/reducer/actions/RangesActions.ts [hunk 1]
```diff
@@ -375,18 +375,58 @@ function handleUpdateRange(draft: Draft<State>, action) {
...
+    draft.toolOptions.data.showRangesIntegrals =
+      !rangeStateInit.showRangesIntegrals;
+  }
 }
 
-function handleShowJGraph(draft: Draft<State>) {
-  draft.toolOptions.data.showJGraph = !draft.toolOptions.data.showJGraph;
+function handleShowJGraph(draft: Draft<State>, action) {
+  const { id } = action.payload;
+  const index = draft.view.ranges.findIndex((r) => r.spectrumID === id);
```
Comment (stropitek) [inline src/component/reducer/actions/RangesActions.ts:418]: ditto

File: src/component/panels/RangesPanel/RangesPanel.tsx [hunk 2]
```diff
@@ -219,23 +223,27 @@ const MemoizedRangesTablePanel = memo(RangesTablePanelInner);
...
     activeTab,
     molecules,
-    toolOptions: {
-      selectedTool,
-      data: { showMultiplicityTrees, showRangesIntegrals, showJGraph },
-    },
+    toolOptions: { selectedTool },
+    view: { ranges: rangeState },
   } = useChartData();
+  // console.log(rangeState);
```
Comment (stropitek) [inline src/component/panels/RangesPanel/RangesPanel.tsx:235]: Please remove this comment

File: src/component/reducer/actions/SpectrumsActions.ts [hunk 3]
```diff
@@ -79,6 +79,22 @@ function handleChangePeaksMarkersVisibility(draft: Draft<State>, data) {
 function handleChangeActiveSpectrum(draft: Draft<State>, activeSpectrum) {
   let refreshDomain = false;
 
+  // Load specturm range state
+  const range = draft.view.ranges.find(
+    (r) => r.spectrumID === activeSpectrum.id,
+  );
+  if (range) {
+    draft.toolOptions.data = {
```
Comment (stropitek) [inline src/component/reducer/actions/SpectrumsActions.ts:87]: @lpatiny 

is `toolOptions.data` still used anywhere?

It seems like it can be removed now

Comment (wadjih-bencheikh18) [reply src/component/reducer/actions/SpectrumsActions.ts:87]: > is `toolOptions.data` still used anywhere and if yes what is its purpose?

I refactored just range tool, `toolOptions.data` still used for other tools states

**Comment by lpatiny** [comment]
The 'jgraph' toggle should also be related to the currently displayed spectrum and currently it does not work correctly

**Comment by wadjih-bencheikh18** [comment]
> it does not work correctly

Fixed, Thank you

**Review by lpatiny — APPROVED** [review]

**Review by stropitek — APPROVED** [review]

**Pull request merged**